internal/runtime/internal/controller/value_cache.go

```diff
@@ -130,107 +135,98 @@ func (vc *valueCache) ExportChangeIndex() int {
-	for blockName, ids := range componentsByBlockName {
-		scope.Variables[blockName] = vc.buildValue(ids, 1)
+	if index == len(id)-1 {
+		delete(m, id[index]) // Remove the component's exports.
+		return nil
 	}
 
-	// Add module arguments to the scope.
-	if len(vc.moduleArguments) > 0 {
-		scope.Variables["argument"] = make(map[string]any)
+	if _, ok := m[id[index]].(map[string]any); !ok {
+		return fmt.Errorf("expected a map but found a value for %q", id[index])
 	}
-	for key, value := range vc.moduleArguments {
-		keyMap := make(map[string]any)
-		keyMap["value"] = value
+	nextM := m[id[index]].(map[string]any)
 
-		switch args := scope.Variables["argument"].(type) {
-		case map[string]any:
-			args[key] = keyMap
-		}
+	err := cleanup(nextM, id, index+1)
+	if err != nil {
+		return err
 	}
 
-	return scope
+	// Delete if the map at this level is empty.
+	// If you only have one Prometheus component and you remove it, it will cleanup the full Prometheus path.
+	// If you have one Prometheus relabel and one Prometheus scrape, and you remove the relabel, it will cleanup the relabel path.
+	if len(nextM) == 0 {
+		delete(m, id[index])
+	}
+	return nil
 }
 
-// buildValue recursively converts the set of user components into a single
-// value. offset is used to determine which element in the userComponentName
-// we're looking at.
-func (vc *valueCache) buildValue(from []ComponentID, offset int) interface{} {
-	// We can't recurse anymore; return the node directly.
-	if len(from) == 1 && offset >= len(from[0]) {
-		name := from[0].String()
-
-		// TODO(rfratto): should we allow arguments to be returned so users can
-		// reference arguments as well as exports?
-		exports, ok := vc.exports[name]
-		if !ok {
-			exports = make(map[string]interface{})
-		}
-		return exports
-	}
+// SyncModuleArgs will remove any cached values for any args no longer in the map.
+func (vc *valueCache) SyncModuleArgs(args map[string]any) {
+	vc.mut.Lock()
+	defer vc.mut.Unlock()
 
-	attrs := make(map[string]interface{})
+	if _, ok := vc.scope.Variables[argumentLabel]; !ok {
+		return
+	}
 
-	// First, partition the components by their label.
-	var componentsByLabel = make(map[string][]ComponentID)
-	for _, id := range from {
-		blockName := id[offset]
-		componentsByLabel[blockName] = append(componentsByLabel[blockName], id)
+	argsMap := vc.scope.Variables[argumentLabel].(map[string]any)
+	for arg := range argsMap {
+		if _, ok := args[arg]; !ok {
+			delete(argsMap, arg)
+		}
 	}
+	if len(argsMap) == 0 {
+		delete(vc.scope.Variables, argumentLabel)
+	}
+}
 
-	// Then, convert each partition into a single value.
-	for label, ids := range componentsByLabel {
-		attrs[label] = vc.buildValue(ids, offset+1)
+// GetScope returns the current scope.
+func (vc *valueCache) GetScope() *vm.Scope {
+	vc.mut.RLock()
+	defer vc.mut.RUnlock()
+	return vm.NewScope(deepCopyMap(vc.scope.Variables))
+}
+
+func deepCopyMap(original map[string]any) map[string]any {
```

Comment:
How does this interact with capsule types? Those are by pointer so will get copied into the new copy?

Reply:
yes that's correct, only the maps are created new but the content of the maps is copied. This was also the case before where the exports values were copied when building the scope